### ISSUE TITLE
ERR_SSL_PROTOCOL_ERROR while serving several domains on a host

### ISSUE THREAD

**Issue by SylleoYr** [caddyserver/caddy#3815]
**Caddy version:** v2.2.1 h1:Q62GWHMtztnvyRU+KPOpw6fNfeCD3SkwH7SfT1Tgt2c= (installed from apt)
**System version:** Ubuntu 20.04

_(I have searched the web and issues, but no answers can fix my problem (it confused me two days), if there is one, please link that to me, thanks a lot. Any insights will be appreciated. Thank you for reading my poor English.)_

emm..., I have a host with several domains (subdomains) refering to, and I want to serve them with different backends, so I write a Caddyfile like the following:

```Caddyfile
{
    http_port 6088
    https_port 6443

    email sylleoyr@xxx.com

}

(secure_headers) {
   header {
        Strict-Transport-Security "max-age=31536000; includeSubDomains; preload"
        X-Frame-Options SAMEORIGIN
        X-Content-Type-Options nosniff
    }
}

(common_encoders) {
    encode zstd gzip
}

(client_tls_auth) {
    tls {
        client_auth {
            mode require_and_verify
            trusted_ca_cert_file /home/user/Apps/Caddy/pki/ca.crt
        }
    }
}

:6088 {
    @ip_direct {
        not host example.com
        not host rs.example.com
        not host rh.example.com
    }

    log {
        output file /var/log/caddy/deny.access.log
    }

    respond @ip_direct "Access Denied." 403 {
        close
    }
}

example.com {
    import common_encoders
    import secure_headers

    log {
        output file /var/log/caddy/access.log
    }

    import client_tls_auth

    handle_path /fd/* {
        reverse_proxy 127.0.0.1:1180
    }

    handle {
        respond "not found" 404 {
            close
        }
    }
}

rs.example.com {
    import common_encoders
    import secure_headers

    log {
        output file /var/log/caddy/rs.access.log
    }
    
    import client_tls_auth

    reverse_proxy 127.0.0.1:1181
}
```

As you see, I binded http port and https port to 6088 and 6443, respectively, and I opened the HSTS. The block under `:6088` is used to prevent accessing through IP and other domains out of my control. Additionally I need a client auth on `example.com` and `rs.example.com`.

But the problem is that when I appended the `rh.example.com` block as the following to my Caddyfile with reloading Caddy, I cannot access `rh.example.com`, the browser showed me `ERR_SSL_PROTOCOL_ERROR`. Then I searched the web and issues, and no answers can fix my problem. A approximate issue is [#2423](https://github.com/caddyserver/caddy/issues/2423) (I met that same problem on my local virtual machine), but I am not very clear about [#2423](https://github.com/caddyserver/caddy/issues/2423) saying (I am a beginner of Web).

```Caddyfile
rh.example.com {
    import common_encoders
    import secure_headers

    log {
        output file /var/log/caddy/rh.access.log
    }

    reverse_proxy 127.0.0.1:1200
}
```

The next I viewed the $HOME directory of Caddy, I found there is no certificate for `rh.example.com` located in `/var/lib/caddy/.local/share/caddy/`, I thought that caddy didn't apply a certificate to `rh.example.como`, so I deleted the whole directory of `/var/lib/caddy/.local/share/caddy/...`, and then reloaded Caddy. Now, I cannot access all my domains. And then I reinstall Caddy, the problem still exists. The following is part of logs printed by `sudo journalctl -u caddy` (the most other logs look same like the following):

```log
Oct 05 18:03:26 ecs007 systemd[1]: Stopped Caddy.
Oct 05 18:03:26 ecs007 systemd[1]: Started Caddy.
Oct 05 18:03:26 ecs007 caddy[22553]: caddy.HomeDir=/var/lib/caddy
Oct 05 18:03:26 ecs007 caddy[22553]: caddy.AppDataDir=/var/lib/caddy/.local/share/caddy
Oct 05 18:03:26 ecs007 caddy[22553]: caddy.AppConfigDir=/var/lib/caddy/.config/caddy
Oct 05 18:03:26 ecs007 caddy[22553]: caddy.ConfigAutosavePath=/var/lib/caddy/.config/caddy/autosave.json
Oct 05 18:03:26 ecs007 caddy[22553]: caddy.Version=v2.2.0
Oct 05 18:03:26 ecs007 caddy[22553]: runtime.GOOS=linux
Oct 05 18:03:26 ecs007 caddy[22553]: runtime.GOARCH=amd64
Oct 05 18:03:26 ecs007 caddy[22553]: runtime.Compiler=gc
Oct 05 18:03:26 ecs007 caddy[22553]: runtime.NumCPU=1
Oct 05 18:03:26 ecs007 caddy[22553]: runtime.GOMAXPROCS=1
Oct 05 18:03:26 ecs007 caddy[22553]: runtime.Version=go1.15.2
Oct 05 18:03:26 ecs007 caddy[22553]: os.Getwd=/
Oct 05 18:03:26 ecs007 caddy[22553]: LANG=en_US.UTF-8
Oct 05 18:03:26 ecs007 caddy[22553]: PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
Oct 05 18:03:26 ecs007 caddy[22553]: HOME=/var/lib/caddy
Oct 05 18:03:26 ecs007 caddy[22553]: LOGNAME=caddy
Oct 05 18:03:26 ecs007 caddy[22553]: USER=caddy
Oct 05 18:03:26 ecs007 caddy[22553]: INVOCATION_ID=860c454123d148c19ec8c0f831f29873
Oct 05 18:03:26 ecs007 caddy[22553]: JOURNAL_STREAM=9:245643
Oct 05 18:03:26 ecs007 caddy[22553]: {"level":"info","ts":1601892206.4964323,"msg":"using provided configuration","config_file":"/etc/caddy/Caddyfile","config_adapter":""}
Oct 05 18:03:26 ecs007 caddy[22553]: {"level":"info","ts":1601892206.5006416,"logger":"admin","msg":"admin endpoint started","address":"tcp/localhost:2019","enforce_origin":false,"origins":["127.0.0.1:2019","localhost:2019","[::1]:2019"]}
Oct 05 18:03:26 ecs007 caddy[22553]: {"level":"info","ts":1601892206.5010579,"logger":"http","msg":"server is listening only on the HTTP port, so no automatic HTTPS will be applied to this server","server_name":"srv0","http_port":80}
Oct 05 18:03:26 ecs007 caddy[22553]: {"level":"info","ts":1601892206.502646,"msg":"autosaved config","file":"/var/lib/caddy/.config/caddy/autosave.json"}
Oct 05 18:03:26 ecs007 caddy[22553]: {"level":"info","ts":1601892206.502795,"msg":"serving initial configuration"}
Oct 05 18:03:26 ecs007 caddy[22553]: {"level":"info","ts":1601892206.5031822,"logger":"tls.cache.maintenance","msg":"started background certificate maintenance","cache":"0xc0004c6690"}
Oct 05 18:03:26 ecs007 caddy[22553]: {"level":"info","ts":1601892206.5040274,"logger":"tls","msg":"cleaned up storage units"}
Oct 06 18:03:26 ecs007 caddy[22553]: {"level":"info","ts":1601978606.5069747,"logger":"tls","msg":"cleaned up storage units"}
Oct 07 18:03:26 ecs007 caddy[22553]: {"level":"info","ts":1602065006.5026484,"logger":"tls","msg":"cleaned up storage units"}
Oct 08 18:03:26 ecs007 caddy[22553]: {"level":"info","ts":1602151406.5024035,"logger":"tls","msg":"cleaned up storage units"}
Oct 09 18:03:26 ecs007 caddy[22553]: {"level":"info","ts":1602237806.5024605,"logger":"tls","msg":"cleaned up storage units"}
Oct 10 18:03:26 ecs007 caddy[22553]: {"level":"info","ts":1602324206.5025315,"logger":"tls","msg":"cleaned up storage units"}
Oct 11 18:03:26 ecs007 caddy[22553]: {"level":"info","ts":1602410606.5026267,"logger":"tls","msg":"cleaned up storage units"}
Oct 12 18:03:26 ecs007 caddy[22553]: {"level":"info","ts":1602497006.5024061,"logger":"tls","msg":"cleaned up storage units"}
Oct 13 18:03:26 ecs007 caddy[22553]: {"level":"info","ts":1602583406.5036342,"logger":"tls","msg":"cleaned up storage units"}
Oct 13 19:57:43 ecs007 caddy[22553]: {"level":"info","ts":1602590263.462933,"logger":"admin.api","msg":"received request","method":"POST","host":"localhost:2019","uri":"/load","remote_addr":"127.0.0.1:36968","headers":{"Accept-Encoding":["gzip"],"Content-Length":["2570"],"Content-Type":["application/json"],"Origin":["localhost:2019"],"User-Agent":["Go-http-client/1.1"]}}
Oct 13 19:57:43 ecs007 caddy[22553]: {"level":"info","ts":1602590263.4637845,"logger":"admin","msg":"admin endpoint started","address":"tcp/localhost:2019","enforce_origin":false,"origins":["localhost:2019","[::1]:2019","127.0.0.1:2019"]}
Oct 13 19:57:43 ecs007 caddy[22553]: {"level":"info","ts":1602590263.4644127,"logger":"http","msg":"enabling automatic HTTP->HTTPS redirects","server_name":"srv0"}
Oct 13 19:57:43 ecs007 caddy[22553]: {"level":"info","ts":1602590263.4644516,"logger":"http","msg":"enabling strict SNI-Host matching because TLS client auth is configured","server_name":"srv0"}
Oct 13 19:57:43 ecs007 caddy[22553]: {"level":"info","ts":1602590263.4657147,"logger":"http","msg":"enabling automatic TLS certificate management","domains":["example.com"]}
Oct 13 19:57:43 ecs007 caddy[22553]: {"level":"info","ts":1602590263.481567,"logger":"tls.cache.maintenance","msg":"started background certificate maintenance","cache":"0xc0004c6230"}
Oct 13 19:57:53 ecs007 caddy[22553]: {"level":"error","ts":1602590273.481698,"logger":"admin","msg":"stopping current admin endpoint","error":"shutting down admin server: context deadline exceeded"}
Oct 13 19:58:13 ecs007 caddy[22553]: {"level":"warn","ts":1602590293.4844923,"logger":"tls","msg":"stapling OCSP","error":"no OCSP stapling for [example.com]: making OCSP request: Post \"http://ocsp.int-x3.letsencrypt.org\": dial tcp 104.244.46.185:80: i/o timeout"}
Oct 13 19:58:13 ecs007 caddy[22553]: {"level":"info","ts":1602590293.9847488,"logger":"tls.cache.maintenance","msg":"stopped background certificate maintenance","cache":"0xc0004c6690"}
Oct 13 19:58:13 ecs007 caddy[22553]: {"level":"info","ts":1602590293.9849968,"msg":"autosaved config","file":"/var/lib/caddy/.config/caddy/autosave.json"}
Oct 13 19:58:13 ecs007 caddy[22553]: {"level":"info","ts":1602590293.9850087,"logger":"admin.api","msg":"load complete"}
Oct 14 07:57:43 ecs007 caddy[22553]: {"level":"info","ts":1602633463.4684415,"logger":"tls","msg":"cleaned up storage units"}
Oct 14 10:04:21 ecs007 caddy[22553]: {"level":"info","ts":1602641061.0313418,"logger":"admin.api","msg":"received request","method":"POST","host":"localhost:2019","uri":"/load","remote_addr":"127.0.0.1:37904","headers":{"Accept-Encoding":["gzip"],"Content-Length":["2811"],"Content-Type":["applica......

Oct 22 18:05:21 ecs007 caddy[22553]: {"level":"error","ts":1603361121.049071,"logger":"tls.issuance.acme.acme_client","msg":"challenge failed","identifier":"rh.example.com","challenge_type":"tls-alpn-01","status_code":400,"problem_type":"urn:ietf:params:acme:error:connection","error":"Timeout during connect (likely firewall problem)"}
Oct 22 18:05:21 ecs007 caddy[22553]: {"level":"error","ts":1603361121.0491273,"logger":"tls.issuance.acme.acme_client","msg":"validating authorization","identifier":"rh.example.com","error":"authorization failed: HTTP 400 urn:ietf:params:acme:error:connection - Timeout during connect (likely firewall problem)","order":"https://acme-staging-v02.api.letsencrypt.org/acme/order/16240865/170971136","attempt":1,"max_attempts":3}
Oct 22 18:05:22 ecs007 caddy[22553]: {"level":"info","ts":1603361122.4539602,"logger":"tls.issuance.acme.acme_client","msg":"trying to solve challenge","identifier":"rh.example.com","challenge_type":"http-01","ca":"https://acme-staging-v02.api.letsencrypt.org/directory"}
Oct 22 18:05:32 ecs007 caddy[22553]: {"level":"error","ts":1603361132.8468614,"logger":"tls.issuance.acme.acme_client","msg":"challenge failed","identifier":"rh.example.com","challenge_type":"http-01","status_code":400,"problem_type":"urn:ietf:params:acme:error:connection","error":"Fetching http://rh.example.com/.well-known/acme-challenge/-lyAORfXkyTIuyuJ9I2yS8iAC4ZnC2sX_W9ukSKum0Q: Timeout during connect (likely firewall problem)"}
Oct 22 18:05:32 ecs007 caddy[22553]: {"level":"error","ts":1603361132.8469493,"logger":"tls.issuance.acme.acme_client","msg":"validating authorization","identifier":"rh.example.com","error":"authorization failed: HTTP 400 urn:ietf:params:acme:error:connection - Fetching http://rh.example.com/.well-known/acme-challenge/-lyAORfXkyTIuyuJ9I2yS8iAC4ZnC2sX_W9ukSKum0Q: Timeout during connect (likely firewall problem)","order":"https://acme-staging-v02.api.letsencrypt.org/acme/order/16240865/170971183","attempt":2,"max_attempts":3}
Oct 22 18:05:34 ecs007 caddy[22553]: {"level":"error","ts":1603361134.4557478,"logger":"tls.obtain","msg":"will retry","error":"[rh.example.com] Obtain: [rh.example.com] solving challenges: rh.example.com: no solvers available for remaining challenges (configured=[http-01 tls-alpn-01] offered=[http-01 dns-01 tls-alpn-01] remaining=[dns-01]) (order=https://acme-staging-v02.api.letsencrypt.org/acme/order/16240865/170971228) (ca=https://acme-staging-v02.api.letsencrypt.org/directory)","attempt":7,"retrying_in":1200,"elapsed":2589.702481056,"max_duration":2592000}
Oct 22 18:08:42 ecs007 caddy[22553]: {"level":"info","ts":1603361322.6280196,"logger":"admin.api","msg":"received request","method":"POST","host":"localhost:2019","uri":"/load","remote_addr":"127.0.0.1:52298","headers":{"Accept-Encoding":["gzip"],"Content-Length":["5447"],"Content-Type":["application/json"],"Origin":["localhost:2019"],"User-Agent":["Go-http-client/1.1"]}}
Oct 22 18:08:42 ecs007 caddy[22553]: {"level":"info","ts":1603361322.6305428,"logger":"admin","msg":"admin endpoint started","address":"tcp/localhost:2019","enforce_origin":false,"origins":["localhost:2019","[::1]:2019","127.0.0.1:2019"]}
Oct 22 18:08:42 ecs007 caddy[22553]: {"level":"info","ts":1603361322.6317277,"logger":"http","msg":"server is listening only on the HTTP port, so no automatic HTTPS will be applied to this server","server_name":"srv0","http_port":6088}
Oct 22 18:08:42 ecs007 caddy[22553]: {"level":"info","ts":1603361322.6322782,"logger":"http","msg":"enabling automatic HTTP->HTTPS redirects","server_name":"srv1"}
Oct 22 18:08:42 ecs007 caddy[22553]: {"level":"warn","ts":1603361322.6324964,"logger":"http","msg":"user server is listening on same interface as automatic HTTP->HTTPS redirects; user-configured routes might override these redirects","server_name":"srv0","interface":"tcp/:6088"}
Oct 22 18:08:42 ecs007 caddy[22553]: {"level":"info","ts":1603361322.632694,"logger":"http","msg":"enabling strict SNI-Host matching because TLS client auth is configured","server_name":"srv1"}
Oct 22 18:08:42 ecs007 caddy[22553]: {"level":"info","ts":1603361322.6338043,"logger":"http","msg":"enabling automatic TLS certificate management","domains":["rs.example.com","example.com","rh.example.com"]}
Oct 22 18:08:42 ecs007 caddy[22553]: {"level":"info","ts":1603361322.631989,"logger":"tls.cache.maintenance","msg":"started background certificate maintenance","cache":"0xc00013e4d0"}

Oct 22 18:09:58 ecs007 caddy[22553]: {"level":"error","ts":1603361398.2071178,"logger":"tls.obtain","msg":"will retry","error":"[rh.example.com] Obtain: [rh.example.com] creating new order: request to https://acme-v02.api.letsencrypt.org/acme/new-order failed after 1 attempts: HTTP 429 urn:ietf:params:acme:error:rateLimited - Error creating new order :: too many failed authorizations recently: see https://letsencrypt.org/docs/rate-limits/ (ca=https://acme-v02.api.letsencrypt.org/directory)","attempt":1,"retrying_in":60,"elapsed":13.565326114,"max_duration":2592000}
Oct 22 18:10:59 ecs007 caddy[22553]: {"level":"info","ts":1603361459.8594837,"logger":"tls.issuance.acme.acme_client","msg":"trying to solve challenge","identifier":"rh.example.com","challenge_type":"tls-alpn-01","ca":"https://acme-staging-v02.api.letsencrypt.org/directory"}
Oct 22 18:11:10 ecs007 caddy[22553]: {"level":"error","ts":1603361470.7122633,"logger":"tls.issuance.acme.acme_client","msg":"challenge failed","identifier":"rh.example.com","challenge_type":"tls-alpn-01","status_code":400,"problem_type":"urn:ietf:params:acme:error:connection","error":"Timeout during connect (likely firewall problem)"}
Oct 22 18:11:10 ecs007 caddy[22553]: {"level":"error","ts":1603361470.7123053,"logger":"tls.issuance.acme.acme_client","msg":"validating authorization","identifier":"rh.example.com","error":"authorization failed: HTTP 400 urn:ietf:params:acme:error:connection - Timeout during connect (likely firewall problem)","order":"https://acme-staging-v02.api.letsencrypt.org/acme/order/16240865/170973273","attempt":1,"max_attempts":3}
Oct 22 18:11:12 ecs007 caddy[22553]: {"level":"info","ts":1603361472.1185265,"logger":"tls.issuance.acme.acme_client","msg":"trying to solve challenge","identifier":"rh.example.com","challenge_type":"http-01","ca":"https://acme-staging-v02.api.letsencrypt.org/directory"}
Oct 22 18:11:23 ecs007 caddy[22553]: {"level":"error","ts":1603361483.0228465,"logger":"tls.issuance.acme.acme_client","msg":"challenge failed","identifier":"rh.example.com","challenge_type":"http-01","status_code":400,"problem_type":"urn:ietf:params:acme:error:connection","error":"Fetching http://rh.example.com/.well-known/acme-challenge/1H9O9mrTQ9yz6rCZP3MzODaz6o42gTRASPhnMrPIeus: Timeout during connect (likely firewall problem)"}
Oct 22 18:11:23 ecs007 caddy[22553]: {"level":"error","ts":1603361483.0229287,"logger":"tls.issuance.acme.acme_client","msg":"validating authorization","identifier":"rh.example.com","error":"authorization failed: HTTP 400 urn:ietf:params:acme:error:connection - Fetching http://rh.example.com/.well-known/acme-challenge/1H9O9mrTQ9yz6rCZP3MzODaz6o42gTRASPhnMrPIeus: Timeout during connect (likely firewall problem)","order":"https://acme-staging-v02.api.letsencrypt.org/acme/order/16240865/170973323","attempt":2,"max_attempts":3}
Oct 22 18:11:24 ecs007 caddy[22553]: {"level":"error","ts":1603361484.6181026,"logger":"tls.obtain","msg":"will retry","error":"[rh.example.com] Obtain: [rh.example.com] solving challenges: rh.example.com: no solvers available for remaining challenges (configured=[http-01 tls-alpn-01] offered=[http-01 dns-01 tls-alpn-01] remaining=[dns-01]) (order=https://acme-staging-v02.api.letsencrypt.org/acme/order/16240865/170973426) (ca=https://acme-staging-v02.api.letsencrypt.org/directory)","attempt":2,"retrying_in":120,"elapsed":99.976310809,"max_duration":2592000}
Oct 22 18:12:42 ecs007 caddy[22553]: {"level":"info","ts":1603361562.542255,"logger":"admin.api","msg":"received request","method":"POST","host":"localhost:2019","uri":"/load","remote_addr":"127.0.0.1:52324","headers":{"Accept-Encoding":["gzip"],"Content-Length":["5447"],"Content-Type":["application/json"],"Origin":["localhost:2019"],"User-Agent":["Go-http-client/1.1"]}}
Oct 22 18:12:42 ecs007 caddy[22553]: {"level":"info","ts":1603361562.5436397,"logger":"admin.api","msg":"config is unchanged"}
Oct 22 18:12:42 ecs007 caddy[22553]: {"level":"info","ts":1603361562.5436573,"logger":"admin.api","msg":"load complete"}
Oct 22 18:13:26 ecs007 caddy[22553]: {"level":"info","ts":1603361606.4343622,"logger":"tls.issuance.acme.acme_client","msg":"trying to solve challenge","identifier":"rh.example.com","challenge_type":"tls-alpn-01","ca":"https://acme-staging-v02.api.letsencrypt.org/directory"}
Oct 22 18:13:37 ecs007 caddy[22553]: {"level":"error","ts":1603361617.1557324,"logger":"tls.issuance.acme.acme_client","msg":"challenge failed","identifier":"rh.example.com","challenge_type":"tls-alpn-01","status_code":400,"problem_type":"urn:ietf:params:acme:error:connection","error":"Timeout during connect (likely firewall problem)"}
```

**Comment by francislavoie** [comment]
Are you forwarding port 80 and 443 to 6088 and 6443 respectively? Let's Encrypt will _only_ attempt to connect on port 80 for the HTTP challenge, and 443 for the ALPN challenge. So if you aren't making it publicly accessible on those ports, ACME will fail. https://caddyserver.com/docs/automatic-https#acme-challenges

**Comment by SylleoYr** [comment]
@francislavoie I have fixed this problem on my host, the reason is caused by my host provider and the `:6088` block. The provider blocked my 80 port by default (for some reasons), thus that made the HTTP challenge failed. And when I reopened 80 port, the `:6088` block influenced the process of HTTP challenge again, it denied the message from Let's Encrypt, which made the validation failed.

Thus one of the solution is adding the `:6088` block after the challenge process, but I think the best way is to use DNS challenge (I choose this one). Now I use [acme.sh](https://github.com/Neilpang/acme.sh) to manage my certificates (because Caddy doesn't built-in this function for my domain name service provider).

Thanks for you look into this issue very much 🥰

---

But there are still some other questions I have found today, and I opened a new issue #3817 .
